crates/ra_ide_api/src/goto_definition.rs

```diff
@@ -126,19 +127,42 @@ pub(crate) fn reference_definition(
             None => {
                 // If we failed to resolve then check associated items
                 if let Some(function) = function {
-                    // Should we do this above and then grab path from the PathExpr?
+                    // Resolve associated item for path expressions
                     if let Some(path_expr) =
                         name_ref.syntax().ancestors().find_map(ast::PathExpr::cast)
                     {
                         let infer_result = function.infer(db);
                         let source_map = function.body_source_map(db);
-                        let expr = ast::Expr::cast(path_expr.syntax()).unwrap();
 
-                        if let Some(res) = source_map
-                            .node_expr(expr)
-                            .and_then(|it| infer_result.assoc_resolutions_for_expr(it.into()))
-                        {
-                            return Exact(NavigationTarget::from_impl_item(db, res));
+                        if let Some(expr) = ast::Expr::cast(path_expr.syntax()) {
+                            if let Some(res) = source_map
+                                .node_expr(expr)
+                                .and_then(|it| infer_result.assoc_resolutions_for_expr(it.into()))
+                            {
+                                return Exact(NavigationTarget::from_impl_item(db, res));
+                            }
+                        }
+                    }
+
+                    // Resolve associated item for path patterns
+                    if let Some(path_pat) =
+                        name_ref.syntax().ancestors().find_map(ast::PathPat::cast)
+                    {
+                        let infer_result = function.infer(db);
+
+                        if let Some(p) = path_pat.path().and_then(hir::Path::from_ast) {
+                            if let Some(pat_id) =
+                                function.body(db).pats().find_map(|(pat_id, pat)| match pat {
```

Comment:
I think this should go via `source_map`, like the expressions above, rather than by iterating all the patterns and comparing paths: the same path might have different meaning in different patterns.

Reply:
I was wondering that. Didn't realize I could use  the `From` trait.

Reply:
Yeah, I've added `From` impls just a couple of days ago. Before that, `Pat::cast(path_pat.syntax()).unwrap()` was possible.